File: tests/crypto/test_crypto.js
/**
 * Crypto library tests — round-trip verification for all modules.
 *
 * Run: node tests/crypto/test_crypto.js
 *
 * Co-Authored-By: Tomaso Nervo (@nervo) <nervo@mindprotocol.ai>
 */

'use strict';

const fs = require('fs');
const path = require('path');
const os = require('os');
const crypto = require('../../lib/crypto');

let passed = 0;
let failed = 0;

function assert(condition, label) {
  if (condition) {
    console.log(`  PASS  ${label}`);
    passed++;
  } else {
    console.error(`  FAIL  ${label}`);
    failed++;
  }
}

async function testSpaceKeyRoundTrip() {
  console.log('\n--- Space Key: round-trip encrypt/decrypt ---');

  const { key } = crypto.generateSpaceKey();
  assert(Buffer.isBuffer(key) && key.length === 32, 'generateSpaceKey returns 32-byte Buffer');

  const plaintext = 'The graph breathes. Nodes pulse. Edges carry energy like blood.';
  const encrypted = crypto.encryptContent(plaintext, key);
  assert(typeof encrypted === 'string', 'encryptContent returns a string');
  assert(encrypted.split(':').length === 3, 'ciphertext has 3 colon-separated parts');

  const decrypted = crypto.decryptContent(encrypted, key);
  assert(decrypted === plaintext, 'decryptContent recovers original plaintext');

  // Encrypt again — IV must differ (different ciphertext)
  const encrypted2 = crypto.encryptContent(plaintext, key);
  assert(encrypted2 !== encrypted, 'two encryptions of same plaintext produce different ciphertext (fresh IV)');

  // Decrypt second ciphertext
  const decrypted2 = crypto.decryptContent(encrypted2, key);
  assert(decrypted2 === plaintext, 'second ciphertext also decrypts correctly');
}

function testIsEncrypted() {
  console.log('\n--- isEncrypted detection ---');

  const { key } = crypto.generateSpaceKey();
  const ct = crypto.encryptContent('test', key);

  assert(crypto.isEncrypted(ct) === true, 'isEncrypted returns true for ciphertext');
  assert(crypto.isEncrypted('hello world') === false, 'isEncrypted returns false for plaintext');
  assert(crypto.isEncrypted('') === false, 'isEncrypted returns false for empty string');
  assert(crypto.isEncrypted('a:b') === false, 'isEncrypted returns false for two-part string');
  assert(crypto.isEncrypted(123) === false, 'isEncrypted returns false for non-string');
  assert(crypto.isEncrypted('abc:def:ghi') === true, 'isEncrypted returns true for valid-looking three-part base64');
}

async function testActorKeysRoundTrip() {
  console.log('\n--- Actor Keys: generate / store / load ---');

  const kp = await crypto.generateActorKeyPair();
  assert(Buffer.isBuffer(kp.publicKey) && kp.publicKey.length === 32, 'publicKey is 32-byte Buffer');
  assert(Buffer.isBuffer(kp.privateKey) && kp.privateKey.length === 32, 'privateKey is 32-byte Buffer');

  // Store to temp directory
  const tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), 'mind-crypto-test-'));
  await crypto.storeActorKeys(tmpDir, kp);

  const pubFile = path.join(tmpDir, 'public_key.b64');
  const privFile = path.join(tmpDir, 'private_key.b64');
  assert(fs.existsSync(pubFile), 'public_key.b64 written');
  assert(fs.existsSync(privFile), 'private_key.b64 written');

  // Load back
  const loaded = await crypto.loadActorKeys(tmpDir);
  assert(loaded.publicKey.equals(kp.publicKey), 'loaded publicKey matches original');
  assert(loaded.privateKey.equals(kp.privateKey), 'loaded privateKey matches original');

  // Cleanup
  fs.rmSync(tmpDir, { recursive: true, force: true });
}

async function testKeyExchangeRoundTrip() {
  console.log('\n--- Key Exchange: wrap / unwrap space key ---');

  // Generate a space key and an actor key pair
  const { key: spaceKey } = crypto.generateSpaceKey();
  const actor = await crypto.generateActorKeyPair();

  // Encrypt the space key for the actor
  const wrapped = await crypto.encryptSpaceKeyForActor(spaceKey, actor.publicKey);
  assert(typeof wrapped === 'string', 'encryptSpaceKeyForActor returns a base64 string');
  assert(!wrapped.includes(':'), 'sealed box is a single base64 blob (no colons)');

  // Decrypt the space key
  const unwrapped = await crypto.decryptSpaceKeyForActor(wrapped, actor.publicKey, actor.privateKey);
  assert(Buffer.isBuffer(unwrapped), 'decryptSpaceKeyForActor returns a Buffer');
  assert(unwrapped.equals(spaceKey), 'unwrapped space key matches original');

  // Wrong key pair should fail
  const other = await crypto.generateActorKeyPair();
  let threw = false;
  try {
    await crypto.decryptSpaceKeyForActor(wrapped, other.publicKey, other.privateKey);
  } catch (e) {
    threw = true;
  }
  assert(threw, 'decryption with wrong key pair throws an error');
}

async function testErrorHandling() {
  console.log('\n--- Error handling ---');

  let threw;

  // encryptContent with wrong key size
  threw = false;
  try { crypto.encryptContent('test', Buffer.alloc(16)); } catch (e) { threw = true; }
  assert(threw, 'encryptContent rejects 16-byte key');

  // decryptContent with corrupted ciphertext
  threw = false;
  const { key } = crypto.generateSpaceKey();
  try { crypto.decryptContent('bad:data:here', key); } catch (e) { threw = true; }
  assert(threw, 'decryptContent rejects corrupted ciphertext');

  // encryptSpaceKeyForActor with wrong key size
  threw = false;
  try { await crypto.encryptSpaceKeyForActor(Buffer.alloc(16), Buffer.alloc(32)); } catch (e) { threw = true; }
  assert(threw, 'encryptSpaceKeyForActor rejects 16-byte space key');

  // loadActorKeys from non-existent directory
  threw = false;
  try { await crypto.loadActorKeys('/tmp/nonexistent-mind-crypto-test-xyz'); } catch (e) { threw = true; }
  assert(threw, 'loadActorKeys rejects missing directory');
}

// --- Run all tests ---

(async () => {
  console.log('Mind Protocol Crypto — Test Suite');
  console.log('==================================');

  await testSpaceKeyRoundTrip();
  testIsEncrypted();
  await testActorKeysRoundTrip();
  await testKeyExchangeRoundTrip();
  await testErrorHandling();

  console.log('\n==================================');
  console.log(`Results: ${passed} passed, ${failed} failed`);

  if (failed > 0) {
    process.exit(1);
  } else {
    console.log('All tests passed.');
  }
})();

File: tests/crypto/test_crypto.py
"""
Mind Protocol Crypto — Python test suite.

Round-trip verification for all crypto modules:
  1. Space key encrypt/decrypt
  2. Actor key pair generate/store/load
  3. Key exchange (wrap/unwrap space key)
  4. is_encrypted detection

Run: python -m pytest tests/crypto/test_crypto.py -v
  or: python tests/crypto/test_crypto.py

Co-Authored-By: Tomaso Nervo (@nervo) <nervo@mindprotocol.ai>
"""

import logging
import os
import sys
import tempfile
import shutil

logger = logging.getLogger(__name__)

# Ensure the repo root is on sys.path so `python.crypto` resolves
REPO_ROOT = os.path.abspath(os.path.join(os.path.dirname(__file__), "..", ".."))
if REPO_ROOT not in sys.path:
    sys.path.insert(0, REPO_ROOT)

from python.crypto import (
    generate_space_key,
    encrypt_content,
    decrypt_content,
    is_encrypted,
    generate_actor_key_pair,
    store_actor_keys,
    load_actor_keys,
    encrypt_space_key_for_actor,
    decrypt_space_key_for_actor,
)


# ---------------------------------------------------------------------------
# 1. Space key: round-trip encrypt / decrypt
# ---------------------------------------------------------------------------

def test_generate_space_key():
    key = generate_space_key()
    assert isinstance(key, bytes), "generate_space_key must return bytes"
    assert len(key) == 32, "space key must be 32 bytes"


def test_encrypt_decrypt_round_trip():
    key = generate_space_key()
    plaintext = "The graph breathes. Nodes pulse. Edges carry energy like blood."

    encrypted = encrypt_content(plaintext, key)
    assert isinstance(encrypted, str), "encrypt_content must return a string"
    assert encrypted.count(":") == 2, "ciphertext must have 3 colon-separated parts"

    decrypted = decrypt_content(encrypted, key)
    assert decrypted == plaintext, "decrypt must recover original plaintext"


def test_fresh_iv_each_call():
    key = generate_space_key()
    plaintext = "same input"
    ct1 = encrypt_content(plaintext, key)
    ct2 = encrypt_content(plaintext, key)
    assert ct1 != ct2, "two encryptions of the same plaintext must produce different ciphertext (fresh IV)"

    # Both must still decrypt correctly
    assert decrypt_content(ct1, key) == plaintext
    assert decrypt_content(ct2, key) == plaintext


def test_encrypt_wrong_key_size():
    try:
        encrypt_content("test", b"\x00" * 16)
        assert False, "should have raised ValueError"
    except ValueError:
        pass


def test_decrypt_corrupted_ciphertext():
    key = generate_space_key()
    try:
        decrypt_content("bad:data:here", key)
        assert False, "should have raised an exception"
    except Exception as e:
        logger.debug("Expected failure decrypting corrupted ciphertext: %s", e)


def test_decrypt_wrong_key():
    key1 = generate_space_key()
    key2 = generate_space_key()
    ct = encrypt_content("secret", key1)
    try:
        decrypt_content(ct, key2)
        assert False, "should have raised an exception"
    except Exception as e:
        logger.debug("Expected failure decrypting with wrong key: %s", e)


# ---------------------------------------------------------------------------
# 2. is_encrypted detection
# ---------------------------------------------------------------------------

def test_is_encrypted_true_for_ciphertext():
    key = generate_space_key()
    ct = encrypt_content("test", key)
    assert is_encrypted(ct) is True, "is_encrypted must return True for ciphertext"


def test_is_encrypted_false_for_plaintext():
    assert is_encrypted("hello world") is False
    assert is_encrypted("") is False


def test_is_encrypted_false_for_two_parts():
    assert is_encrypted("a:b") is False


def test_is_encrypted_false_for_non_string():
    assert is_encrypted(123) is False  # type: ignore[arg-type]


def test_is_encrypted_true_for_valid_looking_base64():
    # Matches JS behavior: three base64 segments = True
    assert is_encrypted("abc:def:ghi") is True


# ---------------------------------------------------------------------------
# 3. Actor keys: generate / store / load
# ---------------------------------------------------------------------------

def test_generate_actor_key_pair():
    kp = generate_actor_key_pair()
    assert "public_key" in kp and "private_key" in kp
    assert isinstance(kp["public_key"], bytes) and len(kp["public_key"]) == 32
    assert isinstance(kp["private_key"], bytes) and len(kp["private_key"]) == 32


def test_store_and_load_actor_keys():
    kp = generate_actor_key_pair()
    tmp_dir = tempfile.mkdtemp(prefix="mind-crypto-test-")

    try:
        store_actor_keys(tmp_dir, kp)

        pub_path = os.path.join(tmp_dir, "public_key.b64")
        priv_path = os.path.join(tmp_dir, "private_key.b64")
        assert os.path.isfile(pub_path), "public_key.b64 must be written"
        assert os.path.isfile(priv_path), "private_key.b64 must be written"

        loaded = load_actor_keys(tmp_dir)
        assert loaded["public_key"] == kp["public_key"], "loaded public key must match"
        assert loaded["private_key"] == kp["private_key"], "loaded private key must match"
    finally:
        shutil.rmtree(tmp_dir)


def test_load_actor_keys_missing_dir():
    try:
        load_actor_keys("/tmp/nonexistent-mind-crypto-test-xyz")
        assert False, "should have raised an exception"
    except Exception as e:
        logger.debug("Expected failure loading keys from missing dir: %s", e)


# ---------------------------------------------------------------------------
# 4. Key exchange: wrap / unwrap space key
# ---------------------------------------------------------------------------

def test_key_exchange_round_trip():
    space_key = generate_space_key()
    actor = generate_actor_key_pair()

    wrapped = encrypt_space_key_for_actor(space_key, actor["public_key"])
    assert isinstance(wrapped, str), "wrapped key must be a base64 string"
    assert ":" not in wrapped, "sealed box must be a single base64 blob (no colons)"

    unwrapped = decrypt_space_key_for_actor(
        wrapped, actor["public_key"], actor["private_key"]
    )
    assert isinstance(unwrapped, bytes), "unwrapped key must be bytes"
    assert unwrapped == space_key, "unwrapped space key must match original"


def test_key_exchange_wrong_key_pair():
    space_key = generate_space_key()
    actor = generate_actor_key_pair()
    other = generate_actor_key_pair()

    wrapped = encrypt_space_key_for_actor(space_key, actor["public_key"])

    try:
        decrypt_space_key_for_actor(wrapped, other["public_key"], other["private_key"])
        assert False, "should have raised an exception"
    except Exception as e:
        logger.debug("Expected failure decrypting with wrong key pair: %s", e)


# ---------------------------------------------------------------------------
# Runner — so the file works both as pytest and as standalone script
# ---------------------------------------------------------------------------

if __name__ == "__main__":
    import traceback

    tests = [
        test_generate_space_key,
        test_encrypt_decrypt_round_trip,
        test_fresh_iv_each_call,
        test_encrypt_wrong_key_size,
        test_decrypt_corrupted_ciphertext,
        test_decrypt_wrong_key,
        test_is_encrypted_true_for_ciphertext,
        test_is_encrypted_false_for_plaintext,
        test_is_encrypted_false_for_two_parts,
        test_is_encrypted_false_for_non_string,
        test_is_encrypted_true_for_valid_looking_base64,
        test_generate_actor_key_pair,
        test_store_and_load_actor_keys,
        test_load_actor_keys_missing_dir,
        test_key_exchange_round_trip,
        test_key_exchange_wrong_key_pair,
    ]

    passed = 0
    failed = 0

    print("Mind Protocol Crypto — Python Test Suite")
    print("=" * 42)

    for t in tests:
        try:
            t()
            print(f"  PASS  {t.__name__}")
            passed += 1
        except Exception:
            print(f"  FAIL  {t.__name__}")
            traceback.print_exc()
            failed += 1

    print("=" * 42)
    print(f"Results: {passed} passed, {failed} failed")

    if failed > 0:
        sys.exit(1)
    else:
        print("All tests passed.")
